modules/javafx.graphics/src/main/java/com/sun/javafx/css/BinarySerializer.java

```diff
@@ -0,0 +1,186 @@
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import javafx.css.PseudoClass;
+import javafx.css.Selector;
+import javafx.css.StyleConverter;
+import javafx.geometry.NodeOrientation;
+
+import static javafx.geometry.NodeOrientation.LEFT_TO_RIGHT;
+import static javafx.geometry.NodeOrientation.RIGHT_TO_LEFT;
+
+/**
+ * Class which can read and write selectors in a binary format.
+ */
+public class BinarySerializer {
+    private static final int TYPE_SIMPLE = 1;
+    private static final int TYPE_COMPOUND = 2;
+
+    public static Selector read(DataInputStream is, String[] strings) throws IOException {
+        int type = is.readByte();
+
+        if (type == TYPE_SIMPLE) {
+            return readSimpleSelector(is, strings);
+        }
+
+        // Backwards compatible, if it is not TYPE_SIMPLE any other value is considered TYPE_COMPOUND
+        return readCompoundSelector(is, strings);
+    }
+
+    public static void write(Selector selector, DataOutputStream os, StyleConverter.StringStore stringStore) throws IOException {
+        if (selector instanceof SimpleSelector s) {
+            writeSimpleSelector(s, os, stringStore);
+        }
+        else if(selector instanceof CompoundSelector s) {
+            writeCompoundSelector(s, os, stringStore);
+        }
+        else {
+            throw new IllegalStateException("support missing for selector type: " + (selector == null ? "null" : selector.getClass()));
+        }
+    }
+
+    private static SimpleSelector readSimpleSelector(DataInputStream is, String[] strings) throws IOException {
+        String name = strings[is.readShort()];
+        int nStyleClasses = is.readShort();
+        List<String> styleClasses = new ArrayList<>();
+
+        for (int n = 0; n < nStyleClasses; n++) {
+            styleClasses.add(strings[is.readShort()]);
+        }
+
+        String id = strings[is.readShort()];
+        int nPseudoclasses = is.readShort();
+        List<String> pseudoclasses = new ArrayList<>();
+
+        for (int n = 0; n < nPseudoclasses; n++) {
+            pseudoclasses.add(strings[is.readShort()]);
+        }
+
+        return new SimpleSelector(name, styleClasses, pseudoclasses, id);
+    }
+
+    private static CompoundSelector readCompoundSelector(DataInputStream is, String[] strings) throws IOException {
+        int nSelectors = is.readShort();
+        List<SimpleSelector> selectors = new ArrayList<>();
+
+        for (int n = 0; n < nSelectors; n++) {
+
+            /*
+             * An extra byte is part of the binary format containing TYPE_SIMPLE that
+             * isn't strictly needed. However, to remain backwards compatible with the
+             * first (and only) supported binary format, this byte must be skipped over
+             * before attempting to read what is known to be a simple selector.
+             */
+
+            is.readByte();
```

Comment:
should we still check the value and throw an IOE if it is wrong _for security reasons_?

Reply:
I'm not entirely sure what security reasons you may be referring to; we're not instantiating classes by string for example, and Java doesn't suffer from out of bounds memory accesses due to a bad interpretation of a binary format.

However, this code is a bit more lenient than the original.  The original code would deserialize this as a `CompoundSelector` but then subsequently fail with a `ClassCastException` as it assumes it will be a `SimpleSelector`. I could make it fully compatible in this regard, although it is extremely unlikely that there are any bad files out there where this would apply (as the old implementation would never load it).

Anyway, I'm fine with rejecting values that are not `TYPE_SIMPLE`.

Reply:
> Anyway, I'm fine with rejecting values that are not `TYPE_SIMPLE`.

That seems like a good idea for maximum compatibility.